Port timeline event flame chart code to use flame_chart_canvas

kenzieschmoll opened the pull request:
Cleanup:
- Removes old timeline flame chart code and replaces it with an implementation using flame chart canvas.
- Makes flame chart canvas code generic and abstract so we can subclass it for unique implementations. We now have two implementations of `FlameChartCanvas`: `TimelineFlameChart` and `CpuProfileFlameChart`.

jacob314 commented on packages/devtools/lib/src/charts/flame_chart_canvas.dart:
```diff
@@ -33,44 +32,39 @@ const _shadedBackgroundColor =
+const rowHeightWithPadding = flameChartRowHeight + rowPadding;
+const flameChartTop = rowHeightWithPadding;
+const flameChartRowHeight = 25.0;
+const flameChartInset = 70;
```
nit: drive by comment: make these consts all explicitly double even though it can be inferred. For example, the 70 probably should be a double for future consistency and to avoid int vs double asserts in the future particularly if this code is ever run on a Dart VM.

jacob314 commented on packages/devtools/lib/src/profiler/cpu_profile_flame_chart.dart:
```diff
@@ -104,3 +109,98 @@ class CpuFlameChart extends CpuProfilerView {
+      final CpuStackFrame parent = stackFrame.parent;
+      double left;
+      if (parent == null) {
+        left = flameChartInset.toDouble();
```
avoid the toDouble by making the const itself a double.

natebosch commented on packages/devtools/lib/src/charts/flame_chart_canvas.dart:
```diff
@@ -87,75 +81,11 @@ abstract class FlameChart {
-        Colors.black,
-        stackFrame,
-      );
+  double getFlameChartWidth();
```
https://dart.dev/guides/language/effective-dart/design#avoid-starting-a-method-name-with-get

Similar throughout.

natebosch commented on packages/devtools/lib/src/charts/flame_chart_canvas.dart:
```diff
@@ -33,44 +32,39 @@ const _shadedBackgroundColor =
   double flameChartWidth;
   double flameChartHeight;
 
   double get flameChartWidthWithInsets =>
```
https://journal.stuffwithstuff.com/2016/06/16/long-names-are-long/

This repeats known context. How about `widthWithInsets`

Similar throughout with other names repeating `flameChart`

natebosch commented on packages/devtools/lib/src/charts/flame_chart_canvas.dart:
```diff
@@ -395,13 +331,16 @@ class FlameChartRow {
 }
 
-class FlameChartNode {
+typedef DisplayTextProvider<T> = String Function(T);
```
https://dart.dev/guides/language/effective-dart/design#prefer-inline-function-types-over-typedefs

The typedef is shorter than the name - is the name adding enough value?

natebosch commented on packages/devtools/lib/src/charts/flame_chart_canvas.dart:
```diff
@@ -428,17 +367,19 @@ class FlameChartNode {
-  final CpuStackFrame stackFrame;
+  final T data;
+
+  DisplayTextProvider displayTextProvider;
```
Should this be a `DisplayTextProvider<T>` instead of a `DisplayTextProvider<dynamic>`

If we inline the function type, would you want `String Function(T)` or `String Function(dynamic)`?

natebosch commented on packages/devtools/lib/src/charts/flame_chart_canvas.dart:
```diff
@@ -428,17 +367,19 @@ class FlameChartNode {
+  String get text => displayTextProvider(data);
 
-  String get tooltip => stackFrame.toString();
+  String get tooltip => data.toString();
```
[optional] Shorter, and may highlight nicer, as `String get tooltip => '$data';`

natebosch commented:
[Optional] The multiple levels of `extends` make this hard for me to follow. Have you considered refactoring to pass in callbacks rather than override methods? To look at composition?

jacob314 commented on packages/devtools/lib/src/timeline/timeline_flame_chart.dart:
```diff
@@ -0,0 +1,157 @@
+class TimelineFlameChart extends FlameChartCanvas<TimelineFrame> {
+  TimelineFlameChart({
+    @required TimelineFrame data,
+    @required flameChartWidth,
```
specify types for constructor parameters

jacob314 commented on packages/devtools/lib/src/profiler/cpu_profile_tables.dart:
```diff
@@ -82,7 +82,7 @@ class CpuBottomUp extends CpuProfilerView {
   @override
   void rebuildView() {
-    final CpuProfileData data = getProfileData();
+    final CpuProfileData data = profileData();
```
there is a little bit of a confusing mixture of 
profileData() methods and getters. Maybe make them all getters?
In general, if getting the data is slow, a method should be used. Otherwise using a gett is cleaner.

kenzieschmoll replied:
renamed to `profileDataProvider` to make it clear that this is a provider.

kenzieschmoll commented:
Addressed comments in latest commit.

@natebosch thanks for all the Dart style suggestions! Re: extends vs. callbacks - I did consider this, but we have to override 3 different methods (one of which is quite large). So it seemed cleaner to put the functionality in a subclass and extend the generic abstract `FlameChartCanvas` rather than pass in 3 different callbacks.
